crates/eips/src/eip7702/auth_list.rs

```diff
@@ -0,0 +1,165 @@
+        use super::constants::MAGIC;
+
+        #[derive(RlpEncodable)]
+        struct Auth {
+            chain_id: ChainId,
+            nonce: OptionalNonce,
+            address: Address,
+        }
+
+        let mut buf = Vec::new();
+        buf.put_u8(MAGIC);
+
+        Auth { chain_id: self.chain_id, nonce: self.nonce, address: self.address }.encode(&mut buf);
+
+        keccak256(buf)
+    }
+
+    /// Convert to a signed authorization by adding a signature.
+    pub const fn into_signed<S>(self, signature: S) -> SignedAuthorization<S> {
+        SignedAuthorization { inner: self, signature }
+    }
+}
+
+/// A signed EIP-7702 authorization.
+#[derive(Debug, Clone, RlpEncodable, RlpDecodable)]
+pub struct SignedAuthorization<S> {
+    inner: Authorization,
+    signature: S,
+}
+
+impl<S> SignedAuthorization<S> {
+    /// Get the `signature` for the authorization.
+    pub const fn signature(&self) -> &S {
+        &self.signature
+    }
+}
+
+#[cfg(feature = "k256")]
+impl SignedAuthorization<alloy_primitives::Signature> {
+    /// Recover the authority for the authorization.
+    ///
+    /// # Note
+    ///
+    /// Implementers should check that the authority has no code.
+    pub fn recover_authority(&self) -> Result<Address, alloy_primitives::SignatureError> {
+        self.signature.recover_address_from_prehash(&self.inner.signature_hash())
+    }
+}
+
+impl<S> Deref for SignedAuthorization<S> {
+    type Target = Authorization;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+/// An internal wrapper around an `Option<u64>` for optional nonces.
+///
+/// In EIP-7702 the nonce is encoded as a list of either 0 or 1 items, where 0 items means that no
+/// nonce was specified (i.e. `None`). If there is 1 item, this is the same as `Some`.
+///
+/// The wrapper type is used for RLP encoding and decoding.
+#[derive(Default, Debug, Copy, Clone)]
+pub struct OptionalNonce(Option<u64>);
+
+impl OptionalNonce {
+    /// Create a new [`OptionalNonce`]
+    pub const fn new(nonce: Option<u64>) -> Self {
+        Self(nonce)
+    }
+}
+
+impl From<Option<u64>> for OptionalNonce {
+    fn from(value: Option<u64>) -> Self {
+        Self::new(value)
+    }
+}
+
+impl Encodable for OptionalNonce {
+    fn encode(&self, out: &mut dyn BufMut) {
+        match self.0 {
+            Some(nonce) => {
+                Header { list: true, payload_length: nonce.length() }.encode(out);
+                nonce.encode(out);
+            }
+            None => Header { list: true, payload_length: 0 }.encode(out),
+        }
+    }
+}
+
+impl Decodable for OptionalNonce {
+    fn decode(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
+        let mut bytes = Header::decode_bytes(buf, true)?;
+        if bytes.is_empty() {
+            return Ok(Self(None));
+        }
+
+        let payload_view = &mut bytes;
+        let nonce = u64::decode(payload_view)?;
+        if !payload_view.is_empty() {
```

Comment:
wouldn't this error if there are for example more things in left to be decoded?

Reply:
I belive this is correct.

maybe we can a few simple rlp roundtrip tests for Authorization